Add Server Timing middleware

sogamoso opened the pull request:
### Summary

#### What is server timing?
It's a specification that defines how the server can communicate the
user agent performance metrics about the request it is responding to.

Here's the official specification: https://www.w3.org/TR/server-timing/

#### How does it work?
This introduces a new `ServerTiming` middleware only on `development` by
default, this is done using the `config.server_timing` setting.

It works by subscribing to all `ActiveSupport::Notifications` and adding
their duration to the `Server-Timing` header.

#### Why is this useful?
It makes looking at performance metrics in development much easier,
especially when using Chrome dev tools which include the metrics in the
Network Inspector. Here's an example:

![](https://d3vv6lp55qjaqc.cloudfront.net/items/371h2y3B3a0U470j040u/Image%202019-05-15%20at%205.40.37%20PM.png?)

Paired on this with @guilleiguaran

sogamoso committed:
Add Server Timing middleware

What is server timing?
It's an specification that defines how the server can communicate the
user agent performance metrics about the request it is responding to.

Here's the oficinal specification:
https://www.w3.org/TR/server-timing/#dfn-server-timing-header-field

How does it work?
This introduces a new `ServerTiming` middleware only on `development` by
default, this is done using the `config.server_timing` setting.

It works by subscribing to all `ActiveSupport::Notifications` and adding
their duration to the `Server-Timing` header.

Why is this useful?
It makes looking at perfomance metrics in development much easier,
especially when using Chrime devtools which includes the metrics in the
Network Inspector. Here's an example:

![](https://d3vv6lp55qjaqc.cloudfront.net/items/371h2y3B3a0U470j040u/Image%202019-05-15%20at%205.40.37%20PM.png?)

Paired on this with @guilleiguaran

guilleiguaran commented:
FYI, This was originally planned for GSoC but it wasn't completed so I asked to @sebasoga to work on this.

https://github.com/railsgsoc/ideas/wiki/2018-Ideas#server-timing-middleware

guilleiguaran commented:
FYI, This was originally planned for GSoC but it wasn't completed so I asked to @sebasoga to work on this.

https://github.com/railsgsoc/ideas/wiki/2018-Ideas#server-timing-middleware

/cc @gsamokovarov

jeremy commented on actionpack/lib/action_dispatch/middleware/server_timing.rb:
```diff
@@ -0,0 +1,36 @@
+      events = []
+      ActiveSupport::Notifications.subscribe(/.*/) do |*args|
+        events << ActiveSupport::Notifications::Event.new(*args)
+      end
```
Need to unsubscribe as well?

sebasoga replied:
You're right. Just added that 9d4d4558dfa4d4eb35d0ddff5b597cdccb1019c9.

jeremy commented:
See also #36220

rafaelfranca commented:
I think we should be merging efforts on #36220 given it is an older PR and the contributor is still active.

guilleiguaran commented on actionpack/test/dispatch/server_timing_test.rb:
```diff
@@ -0,0 +1,57 @@
+
+    def show
+      ActiveSupport::Notifications.instrument("custom.event") do
+        sleep 0.001
```
Since we really don't care about what we are doing here we can just use `true` instead

sebasoga replied:
Done.

sogamoso committed:
Unsuscribe from ActiveSupport notifications

We only want to be subscribed for the life span of the middleware
execution.

sogamoso committed:
No need to sleep

We don't really care about what's going on in the block as long as it
gets executed.

sebasoga commented:
@guilleiguaran, @jeremy, @rafaelfranca: happy to merge the efforts. I think we both took a different approach so a little bit of direction of how you think the API should look like might help us know what the best way to merge efforts is. 

/cc @itsWill

jonathanhefner commented on actionpack/lib/action_dispatch/middleware/server_timing.rb:
```diff
@@ -0,0 +1,37 @@
+
+      header_info = []
+      events.group_by(&:name).each do |event_name, events_collection|
+        header_info << "#{event_name};dur=#{duration_sum(events_collection)}"
```
This could be `events_collection.sum(&:duration)` instead of `duration_sum(events_collection)`.

sebasoga replied:
Nice suggestions, done in 45c9631614437b32af264c3fea3dad0333222456.

igorkasyanchuk commented:
It looks like a cool feature. Maybe try to backport it to the older version of Rails?

sogamoso committed:
Use sum with symbol to proc instead of inject

sebasoga commented:
@igorkasyanchuk I'd be glad to give that a try once we settle on an API for this.

sogamoso committed:
Make Rubocop happy

Got [this message](https://github.com/rails/rails/commit/45c9631614437b32af264c3fea3dad0333222456/checks?check_suite_id=390902272):
```
Offenses:

actionpack/test/dispatch/server_timing_test.rb:41:3: C: Layout/EmptyLinesAroundAccessModifier: Remove a blank line after private.
  private
  ^^^^^^^

2826 files inspected, 1 offense detected
```

gsamokovarov commented on actionpack/lib/action_dispatch/middleware/server_timing.rb:
```diff
@@ -0,0 +1,31 @@
+      end
+
+      status, headers, body = @app.call(env)
+      ActiveSupport::Notifications.unsubscribe(subscriber)
```
Should we unsubscribe in an `ensure` block, to ensure we don't leak subscriptions if the `@app` crashes? 🤔

sebasoga replied:
@gsamokovarov are you thinking about something along these lines?
```rb
status, headers, body = begin
  @app.call(env)
ensure
  ActiveSupport::Notifications.unsubscribe(subscriber)
end
```

gsamokovarov replied:
Yes, something like this should work.

sebasoga replied:
Done in b15a815417c05b8c564b71f499bb4a5c1d24beef.

simi commented on railties/lib/rails/generators/rails/app/templates/config/environments/development.rb.tt:
```diff
@@ -12,6 +12,9 @@ Rails.application.configure do
   config.consider_all_requests_local = true
 
+  # Enable server timing
+  config.server_timing = true
```
What about to make this default for development (also default for framework defaults "6.1" )? That will make it easier to use (it will kick-in on its own once "6.1" defaults are loaded) for upgraded apps as well.

https://github.com/rails/rails/blob/61306d74759faa2efac3f3e0a454187d27befeee/railties/lib/rails/application/configuration.rb#L159

sebasoga replied:
@simi I think the most sensible is to make this the default on development only.  I agree this would be a nice default for apps upgrading to 6.1+. Let's see what people in the core and committers team think about it.

simi replied:
My idea is to make it default in development only, but without need of explicit configuration entry.

sogamoso committed:
Use #map instead of #each

Avoid instantiating an empty array and inserting items into it by using
`map`.

sogamoso committed:
Ensure we don't leak subscriptions when the app crashes

We want to make sure that if the application during crashes the Server
Timing middleware unsubscribes before continuing.

ahorek commented:
I would like to see this feature in a future version of Rails, what else needs to be done?

n-rodriguez commented:
Hi there! Any news?

sebasoga commented:
@jeremy, @guilleiguaran is there anything I can do to move forward with this?

p8 commented:
@sebastians Could you resolve the merge conflicts and add a changelog entry?

sebastians commented:
@p8 will do.

guilleiguaran commented:
@sebastians sorry for the super-late reply, I've merged this and added the CHANGELOG entry in https://github.com/rails/rails/commit/54707cb7a71fa3677cab35c139df6ca3ad685566

I think we should explain the new configuration variable added to the Configuring Guide: https://guides.rubyonrails.org/configuring.html
